clients/src/main/java/org/apache/kafka/clients/consumer/internals/RequestFuture.java

```diff
@@ -37,44 +38,43 @@
  *
  * @param <T> Return type of the result (Can be Void if there is no response)
  */
-public class RequestFuture<T> {
-
-    private boolean isDone = false;
-    private T value;
-    private RuntimeException exception;
-    private List<RequestFutureListener<T>> listeners = new ArrayList<>();
+public class RequestFuture<T> implements ConsumerNetworkClient.PollCondition {
 
+    private final AtomicBoolean isDone = new AtomicBoolean(false);
+    private final AtomicReference<T> value = new AtomicReference<>();
+    private final AtomicReference<RuntimeException> exception = new AtomicReference<>();
```

Comment:
Good question. Volatile pros are: less memory usage and indirection and least powerful abstraction principle. Then there's the verbosity versus clarity trade-offs. Volatiles are clearly less verbose, but AtomicReference makes it clear that the fields may be accessed from multiple threads.